src/System.ObjectModel/src/System/Collections/ObjectModel/KeyedCollection.cs

```diff
@@ -0,0 +1,285 @@
+            }
+
+            TKey oldKey = GetKeyForItem(item);
+            if (!_comparer.Equals(oldKey, newKey))
+            {
+                if (newKey != null)
+                {
+                    AddKey(newKey, item);
+                }
+
+                if (oldKey != null)
+                {
+                    RemoveKey(oldKey);
+                }
+            }
+        }
+
+        protected override void ClearItems()
+        {
+            base.ClearItems();
+            if (_dict != null)
+            {
+                _dict.Clear();
+            }
+
+            _keyCount = 0;
+        }
+
+        protected abstract TKey GetKeyForItem(TItem item);
+
+        protected override void InsertItem(int index, TItem item)
+        {
+            TKey key = GetKeyForItem(item);
+            if (key != null)
+            {
+                AddKey(key, item);
+            }
+            base.InsertItem(index, item);
+        }
+
+        protected override void RemoveItem(int index)
+        {
+            TKey key = GetKeyForItem(Items[index]);
+            if (key != null)
+            {
+                RemoveKey(key);
+            }
+            base.RemoveItem(index);
+        }
+
+        protected override void SetItem(int index, TItem item)
+        {
+            TKey newKey = GetKeyForItem(item);
+            TKey oldKey = GetKeyForItem(Items[index]);
+
+            if (_comparer.Equals(oldKey, newKey))
+            {
+                if (newKey != null && _dict != null)
+                {
+                    _dict[newKey] = item;
+                }
+            }
+            else
+            {
+                if (newKey != null)
+                {
+                    AddKey(newKey, item);
+                }
+
+                if (oldKey != null)
+                {
+                    RemoveKey(oldKey);
+                }
+            }
+            base.SetItem(index, item);
+        }
+
+        private void AddKey(TKey key, TItem item)
+        {
+            if (_dict != null)
+            {
+                _dict.Add(key, item);
+            }
+            else if (_keyCount == _threshold)
+            {
+                CreateDictionary();
+                _dict.Add(key, item);
+            }
+            else
+            {
+                if (Contains(key))
+                {
+                    throw new ArgumentException(SR.Argument_AddingDuplicate);
+                }
+
+                _keyCount++;
+            }
+        }
+
+        private void CreateDictionary()
+        {
+            _dict = new Dictionary<TKey, TItem>(_comparer);
+            foreach (TItem item in Items)
+            {
+                TKey key = GetKeyForItem(item);
+                if (key != null)
+                {
+                    _dict.Add(key, item);
+                }
+            }
+        }
+
+        private void RemoveKey(TKey key)
+        {
+            Contract.Assert(key != null, "key shouldn't be null!");
```

Comment:
@ellismg recently did a sweep of the corefx repro to replace Contract.Asserts with Debug.Asserts/Fails. Can we do the same update here?